src/cpp/core/file_lock/LinkBasedFileLock.cpp

```diff
@@ -0,0 +1,325 @@
+/*
+ * LinkBasedFileLock.cpp
+ *
+ * Copyright (C) 2009-12 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <core/FileLock.hpp>
+
+#include <errno.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <set>
+#include <vector>
+
+#include <core/SafeConvert.hpp>
+#include <core/Algorithm.hpp>
+#include <core/Thread.hpp>
+#include <core/Error.hpp>
+#include <core/Log.hpp>
+#include <core/FilePath.hpp>
+#include <core/FileSerializer.hpp>
+#include <core/system/System.hpp>
+
+#include <boost/foreach.hpp>
+#include <boost/system/error_code.hpp>
+
+namespace rstudio {
+namespace core {
+
+namespace {
+
+std::string pidString()
+{
+   PidType pid = system::currentProcessId();
+   return safe_convert::numberToString((long) pid);
+}
+
+std::string makeHostName()
+{
+   char buffer[256];
+   int status = ::gethostname(buffer, 255);
+   if (status)
+      LOG_ERROR(systemError(errno, ERROR_LOCATION));
+   return std::string(buffer);
+}
+
+const std::string& hostName()
+{
+   static std::string instance = makeHostName();
+   return instance;
+}
+
+std::string threadId()
+{
+   std::stringstream ss;
+   ss << boost::this_thread::get_id();
+   return ss.str().substr(2);
```

Comment:
I guess it's possible that if it's not printed as hex on some platforms then we'd be losing precision (and thus risking collisions). I'd say to be on the safe side let's keep the 0x prefix (i.e. use the string representation unmodified).